File: Editing/Editing View/Workspace/PhotoEditingCanvasBrushStrokeView.m
//  Created by Geoff Pado on 8/27/19.
//  Copyright © 2019 Cocoatype, LLC. All rights reserved.

#import "PhotoEditingCanvasBrushStrokeView.h"

#import "PhotoEditingCanvasView.h"

@interface PhotoEditingCanvasBrushStrokeView ()

@property (nonatomic, strong) PhotoEditingCanvasView *canvasView;
@property (nonatomic) BOOL canvasViewIsResetting;
@property (nonatomic, nullable, strong) UIBezierPath *currentPath;

@end

@implementation PhotoEditingCanvasBrushStrokeView

@synthesize currentPath;

- (instancetype)init {
    if ((self = [super initWithFrame:CGRectZero])) {
        self.translatesAutoresizingMaskIntoConstraints = NO;

        self.canvasView = [[PhotoEditingCanvasView alloc] init];
        self.canvasView.delegate = self;
        [self addSubview:self.canvasView];

        [NSLayoutConstraint activateConstraints:@[
            [self.canvasView.widthAnchor constraintEqualToAnchor:self.widthAnchor multiplier:1.0],
            [self.canvasView.heightAnchor constraintEqualToAnchor:self.heightAnchor multiplier:1.0],
            [self.canvasView.centerXAnchor constraintEqualToAnchor:self.centerXAnchor],
            [self.canvasView.centerYAnchor constraintEqualToAnchor:self.centerYAnchor]
        ]];
    }

    return self;
}

- (UIColor *)color {
    return self.canvasView.color;
}

- (void)setColor:(UIColor *)color {
    self.canvasView.color = color;
}

- (void)updateToolWithCurrentZoomScale:(CGFloat)currentZoomScale {
    [self.canvasView updateToolWithCurrentZoomScale:currentZoomScale];
}

#pragma mark PKCanvasViewDelegate

- (void)canvasViewDrawingDidChange:(PKCanvasView *)canvasView {
    if (self.canvasViewIsResetting) { return; }

    if (@available(iOS 14.0, *)) {
        self.currentPath = [self pathFromDrawing:self.canvasView.drawing];
    }

    [self sendActionsForControlEvents:UIControlEventTouchUpInside];

    self.canvasViewIsResetting = YES;
    self.canvasView.drawing = [PKDrawing new];
    self.canvasViewIsResetting = NO;
}

#pragma mark Path Manipulation

- (UIBezierPath *)newPath {
    UIBezierPath *newPath = [UIBezierPath bezierPath];
    newPath.lineCapStyle = kCGLineCapRound;
    newPath.lineJoinStyle = kCGLineJoinRound;
    newPath.lineWidth = self.canvasView.currentLineWidth;
    return newPath;
}

- (UIBezierPath *)pathFromDrawing:(PKDrawing *)drawing {
    if (@available(iOS 14.0, *)) {
        NSArray<PKStroke *> *strokes = [drawing strokes];
        PKStroke *currentStroke = [strokes lastObject];
        PKStrokePath *strokePath = [currentStroke path];
        UIBezierPath *bezierPath = [self newPath];
        [bezierPath moveToPoint:[strokePath interpolatedLocationAt:0]];
        for (NSUInteger i = 1; i < [strokePath count]; i++) {
            [bezierPath addLineToPoint:[strokePath interpolatedLocationAt:i]];
        }
        return bezierPath;
    } else {
        return [self newPath];
    }
}

#pragma mark Touch Handling

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (@available(iOS 14.0, *)) { return; }

    UITouch *touch = [touches anyObject];
    if (touch == nil) { return; }

    self.currentPath = [self newPath];
    [self.currentPath moveToPoint:[touch locationInView:self]];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (@available(iOS 14.0, *)) { return; }

    UITouch *touch = [touches anyObject];
    if (touch == nil) { return; }
    [self.currentPath addLineToPoint:[touch locationInView:self]];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (@available(iOS 14.0, *)) { return; }

    [self sendActionsForControlEvents:UIControlEventTouchUpInside];
    self.currentPath = nil;
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (@available(iOS 14.0, *)) { return; }

    self.currentPath = nil;
}

@end

File: Editing/Editing View/Workspace/PhotoEditingCanvasBrushStrokeView.swift
//  Created by Geoff Pado on 7/29/19.
//  Copyright © 2019 Cocoatype, LLC. All rights reserved.

import PencilKit
import UIKit

@available(iOS 13.0, *)
class PhotoEditingCanvasBrushStrokeView: UIControl, PhotoEditingBrushStrokeView, PKCanvasViewDelegate {
    init() {
        super.init(frame: .zero)
        translatesAutoresizingMaskIntoConstraints = false

        canvasView.delegate = self
        addSubview(canvasView)

        NSLayoutConstraint.activate([
            canvasView.widthAnchor.constraint(equalTo: widthAnchor),
            canvasView.heightAnchor.constraint(equalTo: heightAnchor),
            canvasView.centerXAnchor.constraint(equalTo: centerXAnchor),
            canvasView.centerYAnchor.constraint(equalTo: centerYAnchor)
        ])
    }

    func updateTool(currentZoomScale: CGFloat) {
        canvasView.updateTool(currentZoomScale: currentZoomScale)
    }

    // MARK: PKCanvasViewDelegate

    private var canvasViewIsDirty = false

    func canvasViewDidBeginUsingTool(_ canvasView: PKCanvasView) {
        canvasViewIsDirty = true
    }

    func canvasViewDidEndUsingTool(_ canvasView: PKCanvasView) {
        guard canvasViewIsDirty else { return }
        canvasViewIsDirty = false
        canvasView.drawing = PKDrawing()
    }

    // MARK: Path Manipulation

    private(set) var currentPath: UIBezierPath?

    private func newPath() -> UIBezierPath {
        let newPath = UIBezierPath()
        newPath.lineCapStyle = .round
        newPath.lineJoinStyle = .round
        newPath.lineWidth = canvasView.currentLineWidth
        return newPath
    }

    // MARK: Touch Handling

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        currentPath = newPath()
        currentPath?.move(to: touch.location(in: self))
    }

    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        currentPath?.addLine(to: touch.location(in: self))
    }

    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        sendActions(for: .touchUpInside)
        currentPath = nil
    }

    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        currentPath = nil
    }

    // MARK: Boilerplate

    private let canvasView = PhotoEditingCanvasView()

    @available(*, unavailable)
    required init(coder: NSCoder) {
        let className = String(describing: type(of: self))
        fatalError("\(className) does not implement init(coder:)")
    }
}
